File: C++/1.basic_concepts/ClassVsObject.cpp
// File: ClassVsObject.cpp
// Purpose: Demonstrates the concept of Classes and Objects in C++

#include <iostream>
#include <string>
using namespace std;

/*
CLASSES AND OBJECTS IN C++
-------------------------
1. CLASS:
    - A blueprint or template for creating objects
    - Defines properties (attributes) and behaviors (methods)
    - Example: Car class defines what every car should have

2. OBJECT:
    - An instance of a class
    - Contains real data and can perform actions
    - Example: myCar is a specific car with actual values

3. RELATIONSHIP:
    Class  -> Blueprint -> What a car should be
    Object -> Instance  -> Actual car you can drive
*/

// Class Definition
class Car
{
public:
    // Properties (also called attributes or member variables)
    string color;
    string name;

    // Methods (functions that belong to the class)
    void displayInfo()
    {
        cout << "Car Information:" << endl;
        cout << "Name: " << name << endl;
        cout << "Color: " << color << endl;
    }
};

int main()
{
    // Creating an object (instance) of the Car class
    Car myCar;

    // Setting object properties
    myCar.color = "red";
    myCar.name = "Ferrari";

    // Displaying car information using the method
    myCar.displayInfo();

    // Creating another object to show multiple instances
    Car friendsCar;
    friendsCar.color = "blue";
    friendsCar.name = "Toyota";
    friendsCar.displayInfo();

    return 0;
}

File: OOP Concepts/Kotlin/1.basic_concepts/ClassVsObject.kt
package oop_revesion

/*
- Class: A class is used to define or describe something from the real world in a program.
    For example, think of a car. In code, we can create a class called Car to describe what a car is — like
    its color, speed, and type of engine. The class also describes what a car can do, like drive or stop.

- Object: An object is an instance of a class — meaning it’s the actual thing created from the class blueprint.
    If the class is like the definition of a car, then an object is a specific car in real life. For example,
    if we have a class Car, we can create objects like myCar or yourCar,
    each with its own specific details, like "red" for the color or "150 km/h" for the speed.

- Difference: A class defines the general idea of something,
    and an object is the real version of that thing with specific properties.

*/


// defining class
class Example {
    // defining variables/properties for the class
    var name: String = ""
}


fun main() {
    // defining object
    val example = Example()
    example.name = "First Example"
    println(example.name)
}